quickwit/quickwit-search/src/extract_timestamp_range.rs

```diff
@@ -0,0 +1,431 @@
+use std::ops::Bound;
+
+use quickwit_query::query_ast::{
+    BoolQuery, QueryAst, QueryAstTransformer, RangeQuery, TermQuery, TermSetQuery,
+};
+use tantivy::DateTime;
+use tracing::warn;
+
+const NANOS_IN_SEC: i64 = 1_000_000_000;
+
+/// Extract timestamp range in QueryAst.
+///
+/// This can allow MatchAll queries (post timestamp range removal) to be optimized.
+pub(crate) fn extract_start_end_timestamp_from_ast(
+    query_ast: QueryAst,
+    timestamp_field: &str,
+    start_timestamp: &mut Option<i64>,
+    end_timestamp: &mut Option<i64>,
+) -> QueryAst {
```

Comment:
This signature is confusing as it is both transforming an owned input and mutating references.

Reply:
Hmm, do you maybe prefer:

```rust
pub(crate) fn extract_start_end_timestamp_from_ast(
    query_ast: QueryAst,
    timestamp_field: &str,
    start_timestamp: Option<i64>,
    end_timestamp: Option<i64>,
) -> (QueryAst, Option<i64>, Option<i64>) {
```

?

Reply:
I think the gain from making both the boundary refinement and the extraction at once is marginal, and it would make this PR a lot easier to review (and the code easier to read) to keep `refine_start_end_timestamp_from_ast()` as is and add a separate method for the extraction.

Reply:
The problem is with `transform` accepting a value and not borrow of `QueryAst`, meaning I would need to `clone` to keep the old signature of `refine_start_end_timestamp_from_ast` which accepts a `&QueryAst`, are you ok with this `clone`?

Reply:
fixed